commit d3dc88c720fdbf56a64f3990148387649f366aeb
Author: Ingo Franzki <ifranzki@linux.ibm.com>
Date:   Thu Nov 6 15:27:36 2025 +0100

    EP11: Fix unwrapping of attribute bound EC keys
    
    Newer EP11 firmware versions do not allow to get attribute CKA_VALUE_LEN
    for non-symmetric keys, because CKA_VALUE_LEN is not defined for these
    key types. This now causes error CKR_ATTRIBUTE_TYPE_INVALID.
    Older firmware versions ignored that and did not fail if CKA_VALUE_LEN
    was attempted to be retrieved.
    
    Fix this by only attempting to obtain CKA_VALUE_LEN for keys that support
    this attribute.
    
    Signed-off-by: Ingo Franzki <ifranzki@linux.ibm.com>

diff --git a/usr/lib/ep11_stdll/ep11_specific.c b/usr/lib/ep11_stdll/ep11_specific.c
index 1baa5443..4e6944e6 100644
--- a/usr/lib/ep11_stdll/ep11_specific.c
+++ b/usr/lib/ep11_stdll/ep11_specific.c
@@ -2222,17 +2222,26 @@ static CK_RV ab_unwrap_update_template(STDLL_TokData_t * tokdata,
         {CKA_KEY_TYPE,          &template_keytype,  sizeof(template_keytype)},
         {CKA_VALUE_LEN,         &valuelen,          sizeof(valuelen)},
     };
+    CK_ULONG num_attrs = sizeof(attrs) / sizeof(CK_ATTRIBUTE);
     CK_ULONG i;
     CK_ATTRIBUTE *attr;
     CK_BBOOL cktrue = TRUE;
     CK_BYTE *useblob;
     size_t useblob_len;
 
+    switch (keytype) {
+    case CKK_GENERIC_SECRET:
+    case CKK_AES:
+        break;
+    default:
+        num_attrs -= 1; /* No CKA_VALUE_LEN attribute */
+        break;
+    }
+
     RETRY_SESSION_SINGLE_APQN_START(rc, tokdata)
     RETRY_REENC_BLOB_START(tokdata, target_info, obj, blob, blob_len,
                            useblob, useblob_len, rc)
-        rc = dll_m_GetAttributeValue(useblob, useblob_len, attrs,
-                                     sizeof(attrs) / sizeof(CK_ATTRIBUTE),
+        rc = dll_m_GetAttributeValue(useblob, useblob_len, attrs, num_attrs,
                                      target_info->target);
     RETRY_REENC_BLOB_END(tokdata, target_info, useblob, useblob_len, rc)
     RETRY_SESSION_SINGLE_APQN_END(rc, tokdata, session)
